mars/oscar/backends/mars/pool.py

```diff
@@ -878,147 +841,267 @@ async def create(cls, config: Dict) -> "MainActorPool":
-        for address, process in to_stop_processes.items():
-            tasks.append(self._stop_sub_pool(address, process))
-        await asyncio.gather(*tasks)
-
-    def _process_sub_pool_lost(self, address: str):
-        if self._auto_recover in (False, 'process'):
-            # process down, when not auto_recover
-            # or only recover process, remove all created actors
-            self._allocated_actors[address] = dict()
-
-    async def _recover_sub_pool(self, address: str):
-        process_index = self._config.get_process_index(address)
-        # process dead, restart it
-        loop = asyncio.get_running_loop()
-        executor = futures.ThreadPoolExecutor(1)
-        task = loop.run_in_executor(
-            executor, _start_sub_pool_in_process,
-            self._config, process_index,
-            self._subprocess_start_method)
-        await asyncio.sleep(0)
-        self._sub_processes[address] = await task
-
-        if self._auto_recover == 'actor':
-            # need to recover all created actors
-            for _, message in self._allocated_actors[address].values():
-                create_actor_message: CreateActorMessage = message
-                await self.call(address, create_actor_message)
-
-    async def _monitor_sub_pools(self):
-        try:
-            while not self._stopped.is_set():
-                for address in self._sub_processes:
-                    process = self._sub_processes[address]
-                    if not process.is_alive():
-                        if self._on_process_down is not None:
-                            self._on_process_down(self, address)
-                        self._process_sub_pool_lost(address)
-                        if self._auto_recover:
-                            await self._recover_sub_pool(address)
-                            if self._on_process_recover is not None:
-                                self._on_process_recover(self, address)
-
-                # check every half second
-                await asyncio.sleep(.5)
-        except asyncio.CancelledError:
-            # cancelled
-            return
-
     @implements(AbstractActorPool.start)
     async def start(self):
         await super().start()
         if self._monitor_task is None:
-            self._monitor_task = asyncio.create_task(self._monitor_sub_pools())
+            self._monitor_task = asyncio.create_task(self.sub_actor_pool_manager.monitor_sub_pools())
             await asyncio.sleep(0)
 
     @implements(AbstractActorPool.stop)
     async def stop(self):
-        await self._stop_sub_pools()
+        await self.sub_actor_pool_manager.stop_sub_pools()
         await super().stop()
         if self._monitor_task and not self._monitor_task.done():
             await self._monitor_task
             self._monitor_task = None
 
+    @classmethod
+    def get_sub_pool_manager_cls(cls):
+        return MainActorPool.ProcessSubActorPoolManager
+
+    class SubActorPoolManager(ABC):
```

Comment:
If there is a parent class for `ActorPool`, this class may not be necessary.

Reply:
`SubActorPoolManager` is only used for `MainActorPool`, `SuhActorPool` doesn't need this

Reply:
So the parent `MainActorPool` class has some methods to manage subprocesses.